### PR TITLE
First pass at a GenomicsDB API.

### PR DESCRIPTION
**Pull request by nalinigans**
This API is experimental and exists only for querying of variants and variant calls. Most of the internal classes are opaque and the api is loosely based on the pimpl paradigm.
The GenomicsDB class exposes the following methods
- query_variants()
- query_variant_calls()

`query_variants()` returns GenomicsDBResults<genomicsdb_variant_t> that contains Variants and allows for iterating/navigating through the Variants and VariantCalls contained in the Variants.
`query_variant_calls()` is a wrapper around the underlying SingleCellTileDBIterator and allows for registration of a GenomicsDBVariantCallProcessor with callbacks to process on a per cell basis on the client side. See [GenomicsDB-R/genomicsdb_query.cpp](https://github.com/nalinigans/GenomicsDB-R/blob/5ef6a45c388e54e5e5f7fde8d4b8a8e7fbbb8e7f/src/genomicsdb_query.cpp#L164) for an example processor in the GenomicsDB R Bindings.

The GenomicsDB class can be instantiated with workspace, callset_mapping_file, vid_mapping_file and reference_genome specifications. Attributes and segment_size can be optionally specified. The query methods have to be called with at least the array name as an argument.
There is also an alternate GenomicsDB constructor with query and loader json files and a rank number as input. Queries can be called with no arguments as the the array, column and row ranges are deduced from the rank number and the json files.

There are also a bunch of templated utilities to help with interpreting the results -
interval_t get_interval() on variant and variant calls
int64_t get_row() on variant calls
genomic_interval_t get_genomic_interval() on variant and variant calls
std::vector<genomic_field_t> get_genomic_fields() on variant and variant calls.

Note that get_genomic_fields() returns a vector of fields with the field values in a string form for now. This will evolve into a more logical value based on the field type in a later pass.

Also, in this PR, there is a new cmake target to uninstall GenomicsDB - `uninstall-genomicsdb`.

### CODE REVIEW

**Comment by jacmarjorie** [comment]
It would be useful for `query_variants()` to optionally return top-level variant information without the  genotypes field. This should make it easy to associate calls from `query_variant_calls()` back to variant metadata (reference, alternates, etc.) without having to carry along unnecessary, and repetitive genotype information.

File: src/main/cpp/src/api/genomicsdb.cc [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+interval_t GenomicsDB::get_interval(const genomicsdb_variant_call_t* variant_call) {
+  return get_interval_for(TO_VARIANT_CALL(variant_call));
+}
+
+// Template to get the genomic interval from the GenomicsDB array associated with the Variant(Call)
+template<class VariantOrVariantCall>
+genomic_interval_t get_genomic_interval_for(const VariantOrVariantCall* variant_or_variant_call, VidMapper *vid_mapper) {
+  std::string contig_name;
+  int64_t contig_position;
+  vid_mapper->get_contig_location(variant_or_variant_call->get_column_begin(), contig_name,
```
Comment (mlathara) [inline src/main/cpp/src/api/genomicsdb.cc]: we should check the return from vid_mapper->get_contig_location here to ensure that column successfully mapped to contig_name/_position

File: src/main/cpp/src/api/genomicsdb.cc [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+    delete TO_VARIANT_QUERY_CONFIG(m_query_config);
+  }
+  if (m_storage_manager != nullptr) {
+    delete TO_VARIANT_STORAGE_MANAGER(m_storage_manager);
+  }
+}
+
+GenomicsDBVariants GenomicsDB::query_variants(const std::string& array,
+					       genomicsdb_ranges_t column_ranges,
+					       genomicsdb_ranges_t row_ranges) {
```
Comment (mlathara) [inline src/main/cpp/src/api/genomicsdb.cc]: should we also offer a way to pass in query_filter through this route? Or just ask the user to go the json route for that?

File: src/main/cpp/src/api/genomicsdb.cc [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+
+   assert(query_config.is_defined_query_idx_for_known_field_enum(GVCF_END_IDX));
+   auto end_query_idx = query_config.get_query_idx_for_known_field_enum(GVCF_END_IDX);
+   auto end_position = *(reinterpret_cast<const int64_t*>(cell.get_field_ptr_for_query_idx(end_query_idx)));
+
+   assert(m_vid_mapper);
+
+   std::string contig_name;
+   int64_t contig_position;
+   m_vid_mapper->get_contig_location(coords[1], contig_name, contig_position);
```
Comment (mlathara) [inline src/main/cpp/src/api/genomicsdb.cc]: check bool value for get_contig_location?

File: src/test/cpp/src/test_genomicsdb_api.cc [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+   auto variant1 = variants.at(0);
+   REQUIRE(variant == variant1);
+   auto interval = gdb->get_interval(variant);
+   CHECK(interval.first == 12140);
+   CHECK(interval.second == 12294);
+   auto genomic_interval = gdb->get_genomic_interval(variant);
+   CHECK(genomic_interval.contig_name == "1");
+   CHECK(genomic_interval.interval.first == 12141);
+   CHECK(genomic_interval.interval.second == 12295);
+   auto genomic_fields = gdb->get_genomic_fields(array, variant);
```
Comment (mlathara) [inline src/test/cpp/src/test_genomicsdb_api.cc]: So I understand...can you explain why genomic_fields are empty for Variants, but not for VariantCalls? I'm sure it has to do with whatever test array was ingested/imported...but I would think the ref/alt fields for instance would apply to the Variants

File: src/test/cpp/src/test_genomicsdb_api.cc [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+  void process(uint32_t row,
+               genomic_interval_t genomic_interval,
+               std::vector<genomic_field_t> genomic_fields){
+    m_processed_rows++;
+  };
+
+  int m_num_query_intervals = 0;
+  int m_processed_rows = 0;
+};
+
```
Comment (mlathara) [inline src/test/cpp/src/test_genomicsdb_api.cc]: I'm sure there's an endless list of things we could try to test here...but maybe also add a test case that exercises the rank part of this to make sure that if someone does try to throw more threads at this, we know it'll work.

Comment (nalinigans) [reply src/test/cpp/src/test_genomicsdb_api.cc]: Yes, I will add a test with rank.

File: src/test/cpp/src/test_genomicsdb_api.cc (hunk 4 above)
Comment (nalinigans) [reply src/test/cpp/src/test_genomicsdb_api.cc]: Genomic Fields at the top level Variants are common fields from the contained Variant Calls. When the number of Variant Calls is 1, this field is always empty @kgururaj may have more insight here. I can open up an Issue and fix if need be.

File: src/main/cpp/src/api/genomicsdb.cc (hunk 2 above)
Comment (nalinigans) [reply src/main/cpp/src/api/genomicsdb.cc]: Good idea, this was a first pass and I wanted to keep it simple. Just wondering if there is anything else needed for configuring these queries.

**Comment by nalinigans** [comment]
> Looks good to me - some of the structures used have a lot in common with the Protobuf structs. In a future PR, I can help unify some of them.

Yes, we sorely need protobuf, especially to interpret the fields. Look forward to your effort in unifying the api structures with protobuf. For now, I am concentrating my efforts in getting first implementations of R/Python bindings.

**Pull request merged**